connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigStorage.java

```diff
@@ -331,28 +325,18 @@ public void putTaskConfigs(Map<ConnectorTaskId, Map<String, String>> configs) {
             Struct connectConfig = new Struct(TASK_CONFIGURATION_V0);
             connectConfig.put("properties", taskConfigEntry.getValue());
             byte[] serializedConfig = converter.fromConnectData(topic, TASK_CONFIGURATION_V0, connectConfig);
-            configLog.send(TASK_KEY(taskConfigEntry.getKey()), serializedConfig);
+            sendToConfigLog(TASK_KEY(taskConfigEntry.getKey()), serializedConfig, "Failed to write task configuration to Kafka");
         }
 
         // Finally, send the commit to update the number of tasks and apply the new configs, then wait until we read to
         // the end of the log
-        try {
-            // Read to end to ensure all the task configs have been written
-            configLog.readToEnd().get(READ_TO_END_TIMEOUT_MS, TimeUnit.MILLISECONDS);
-
-            // Write all the commit messages
-            for (Map.Entry<String, Integer> taskCountEntry : newTaskCounts.entrySet()) {
-                Struct connectConfig = new Struct(CONNECTOR_TASKS_COMMIT_V0);
-                connectConfig.put("tasks", taskCountEntry.getValue());
-                byte[] serializedConfig = converter.fromConnectData(topic, CONNECTOR_TASKS_COMMIT_V0, connectConfig);
-                configLog.send(COMMIT_TASKS_KEY(taskCountEntry.getKey()), serializedConfig);
-            }
+        // Write all the commit messages
 
-            // Read to end to ensure all the commit messages have been written
-            configLog.readToEnd().get(READ_TO_END_TIMEOUT_MS, TimeUnit.MILLISECONDS);
-        } catch (InterruptedException | ExecutionException | TimeoutException e) {
-            log.error("Failed to write root configuration to Kafka: ", e);
-            throw new ConnectException("Error writing root configuration to Kafka", e);
+        for (Map.Entry<String, Integer> taskCountEntry : newTaskCounts.entrySet()) {
+            Struct connectConfig = new Struct(CONNECTOR_TASKS_COMMIT_V0);
+            connectConfig.put("tasks", taskCountEntry.getValue());
+            byte[] serializedConfig = converter.fromConnectData(topic, CONNECTOR_TASKS_COMMIT_V0, connectConfig);
+            sendToConfigLog(COMMIT_TASKS_KEY(taskCountEntry.getKey()), serializedConfig, "Failed to send task config commit to Kafka");
```

Comment:
Isn't this going to be really expensive if we `sendToConfigLog` for every item in the for loop since that is blocking on a read to the end of the log for *every* element rather than writing them all and *then* reading to the end? I haven't checked the rest of the patch carefully, but in general I tried to be careful when writing this code to invoke `configLog.readToEnd()` only when necessary, so I'm a bit skeptical of adding a utility method like this that we use in many places.